lib/ecto/query/builder.ex

```diff
@@ -1133,21 +1133,24 @@ defmodule Ecto.Query.Builder do
   the query properly, but they will be in their runtime form
   when invoked at runtime.
   """
+  @spec apply_query(Macro.t, Macro.t, Macro.t, Macro.Env.t) :: Macro.t
   def apply_query(query, module, args, env) do
-    query = Macro.expand(query, env)
+    case Macro.expand(query, env) |> unescape_query() do
+      %Query{} = compiletime_query ->
+        apply(module, :apply, [compiletime_query | args])
+        |> escape_query()
 
-    case unescape_query(query) do
-      %Query{} = unescaped ->
-        apply(module, :apply, [unescaped|args]) |> escape_query
-      _ ->
+      runtime_query ->
         quote do
-          query = unquote(query) # Unquote the query for any binding variable
+          # Unquote the query before `module.apply()` for any binding variable.
+          query = unquote(runtime_query)
```

Comment:
If we unquote the unescaped query, it will be invalid AST. We should revert this change. :)

Reply:
Oh, ignore me, we don't actually unescape it in those case.